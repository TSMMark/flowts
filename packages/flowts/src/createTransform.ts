import * as babel from '@babel/core';
import * as prettier from 'prettier';
import recastPlugin from '@flowts/babel-plugin-recast';
import { PluginItem } from '@babel/core';
import { Options } from 'prettier';

/**
 * Helper utility creating function transforming the code given plugins array
 * @param plugins
 * @param prettierConfig
 * @param recast
 */
export function createTransform(
  plugins: PluginItem[],
  { prettierConfig = {} as Options, recast = true } = {}
) {
  const transformPlugins: PluginItem[] = [];
  if (recast) {
    transformPlugins.push(recastPlugin);
  }

  return function transform(code: string) {
    const transformResult = babel.transformSync(code, {
      compact: false,
      babelrc: false,
      configFile: false,
      plugins: [...transformPlugins, ...plugins],
    });

    if (transformResult === null) {
      throw new Error(
        'babel.transformSync returned null, probably configuration issue'
      );
    }
    prettierConfig.parser = 'babel-ts';
    return prettier.format(transformResult.code as string, {
      ...prettierConfig,
      // avoid changing template strings
      embeddedLanguageFormatting: 'off',
    });
  };
}
