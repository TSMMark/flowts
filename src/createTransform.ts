import * as babel from '@babel/core';
import * as prettier from 'prettier';
import recastPlugin from './recastPlugin';
import { PluginItem } from '@babel/core';
import { Options } from 'prettier';

/**
 * Helper utility creating function transforming the code given plugins array
 * @param plugins
 * @param prettierConfig
 * @param recast
 */
export function createTransform(
  plugins: PluginItem[],
  { prettierConfig = {} as Options, recast = true } = {}
) {
  const transformPlugins: PluginItem[] = [];
  if (recast) {
    transformPlugins.push(recastPlugin);
  }

  return function transform(code: string) {
    const transformResult = babel.transformSync(code, {
      babelrc: false,
      plugins: [...transformPlugins, ...plugins],
    });

    if (transformResult === null) {
      throw new Error(
        'babel.transformSync returned null, probably configuration issue'
      );
    }
    prettierConfig.parser = 'typescript';
    return prettier.format(transformResult.code as string, prettierConfig);
  };
}
