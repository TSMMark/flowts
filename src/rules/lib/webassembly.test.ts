import plugin from '../../tsTypesPlugin';
import { createTransform } from '../../createTransform';
const transform = createTransform([plugin]);

describe('lib/webassembly.js', () => {
  describe('globals', () => {
    describe('ExportedFunctionExoticObject', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = ExportedFunctionExoticObject;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new ExportedFunctionExoticObject();
            
            class A1 extends ExportedFunctionExoticObject {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('WebAssembly$Module', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WebAssembly$Module;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new WebAssembly$Module();
            
            class A1 extends WebAssembly$Module {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('WebAssembly$Instance', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WebAssembly$Instance;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new WebAssembly$Instance();
            
            class A1 extends WebAssembly$Instance {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('WebAssembly$Memory', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WebAssembly$Memory;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new WebAssembly$Memory();
            
            class A1 extends WebAssembly$Memory {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('WebAssembly$Table', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WebAssembly$Table;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new WebAssembly$Table();
            
            class A1 extends WebAssembly$Table {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('WebAssembly$CompileError', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WebAssembly$CompileError;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new WebAssembly$CompileError();
            
            class A1 extends WebAssembly$CompileError {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('WebAssembly$LinkError', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WebAssembly$LinkError;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new WebAssembly$LinkError();
            
            class A1 extends WebAssembly$LinkError {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('WebAssembly$RuntimeError', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WebAssembly$RuntimeError;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new WebAssembly$RuntimeError();
            
            class A1 extends WebAssembly$RuntimeError {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('WebAssembly$instantiate', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WebAssembly$instantiate;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class', () => {
        expect(
          transform(`
            new WebAssembly$instantiate();
            
            class A1 extends WebAssembly$instantiate {};
          `)
        ).toMatchSnapshot();
      });

      test('generated - callable', () => {
        expect(
          transform(`
            WebAssembly$instantiate();
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('WebAssembly', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = WebAssembly;
          `)
        ).toMatchSnapshot();
      });
    });
  });

  describe('modules', () => {
    test('has no modules', () => {});
  });
});
