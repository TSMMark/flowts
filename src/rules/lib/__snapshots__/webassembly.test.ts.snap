// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/webassembly.js globals ExportedFunctionExoticObject generated - class 1`] = `
"new ExportedFunctionExoticObject();

class A1 extends ExportedFunctionExoticObject {}
"
`;

exports[`lib/webassembly.js globals ExportedFunctionExoticObject variable 1`] = `
"var a = ExportedFunctionExoticObject;
"
`;

exports[`lib/webassembly.js globals WebAssembly variable 1`] = `
"var a = WebAssembly;
"
`;

exports[`lib/webassembly.js globals WebAssembly$CompileError generated - class 1`] = `
"new WebAssembly$CompileError();

class A1 extends WebAssembly$CompileError {}
"
`;

exports[`lib/webassembly.js globals WebAssembly$CompileError variable 1`] = `
"var a = WebAssembly$CompileError;
"
`;

exports[`lib/webassembly.js globals WebAssembly$Instance generated - class 1`] = `
"new WebAssembly$Instance();

class A1 extends WebAssembly$Instance {}
"
`;

exports[`lib/webassembly.js globals WebAssembly$Instance variable 1`] = `
"var a = WebAssembly$Instance;
"
`;

exports[`lib/webassembly.js globals WebAssembly$LinkError generated - class 1`] = `
"new WebAssembly$LinkError();

class A1 extends WebAssembly$LinkError {}
"
`;

exports[`lib/webassembly.js globals WebAssembly$LinkError variable 1`] = `
"var a = WebAssembly$LinkError;
"
`;

exports[`lib/webassembly.js globals WebAssembly$Memory generated - class 1`] = `
"new WebAssembly$Memory();

class A1 extends WebAssembly$Memory {}
"
`;

exports[`lib/webassembly.js globals WebAssembly$Memory variable 1`] = `
"var a = WebAssembly$Memory;
"
`;

exports[`lib/webassembly.js globals WebAssembly$Module generated - class 1`] = `
"new WebAssembly$Module();

class A1 extends WebAssembly$Module {}
"
`;

exports[`lib/webassembly.js globals WebAssembly$Module variable 1`] = `
"var a = WebAssembly$Module;
"
`;

exports[`lib/webassembly.js globals WebAssembly$RuntimeError generated - class 1`] = `
"new WebAssembly$RuntimeError();

class A1 extends WebAssembly$RuntimeError {}
"
`;

exports[`lib/webassembly.js globals WebAssembly$RuntimeError variable 1`] = `
"var a = WebAssembly$RuntimeError;
"
`;

exports[`lib/webassembly.js globals WebAssembly$Table generated - class 1`] = `
"new WebAssembly$Table();

class A1 extends WebAssembly$Table {}
"
`;

exports[`lib/webassembly.js globals WebAssembly$Table variable 1`] = `
"var a = WebAssembly$Table;
"
`;

exports[`lib/webassembly.js globals WebAssembly$instantiate generated - callable 1`] = `
"WebAssembly$instantiate();
"
`;

exports[`lib/webassembly.js globals WebAssembly$instantiate generated - class 1`] = `
"new WebAssembly$instantiate();

class A1 extends WebAssembly$instantiate {}
"
`;

exports[`lib/webassembly.js globals WebAssembly$instantiate variable 1`] = `
"var a = WebAssembly$instantiate;
"
`;
